File: cpp-std/exer06b01-blocking-queue.cpp
/*
BLOCKING QUEUE IMPLEMENTATION
Version B01: General blocking queues
             Underlying mechanism: Semaphores
*/


#include <iostream>
#include <queue>
#include <string>
#include <chrono>
#include <stdexcept>
#include <thread>
#include <mutex>
#include <semaphore>
using namespace std;



using cntsemaphore = std::counting_semaphore<>;



template <typename T>
class BlockingQueue {

private:
    int capacity;

    cntsemaphore semRemain;
    cntsemaphore semFill;
    std::mutex mut;

    std::queue<T> q;


public:
    BlockingQueue(int capacity) : capacity(capacity), semRemain(capacity), semFill(0) {
        if (this->capacity <= 0)
            throw std::invalid_argument("capacity must be a positive integer");
    }


    void put(const T& value) {
        semRemain.acquire();

        {
            std::unique_lock<std::mutex> lk(mut);
            q.push(value);
        }

        semFill.release();
    }


    T take() {
        T result;
        semFill.acquire();

        {
            std::unique_lock<std::mutex> lk(mut);
            result = q.front();
            q.pop();
        }

        semRemain.release();
        return result;
    }

};



void producer(BlockingQueue<std::string>* blkQueue) {
    auto arr = { "nice", "to", "meet", "you" };

    for (auto&& data : arr) {
        cout << "Producer: " << data << endl;
        blkQueue->put(data);
        cout << "Producer: " << data << "\t\t\t[done]" << endl;
    }
}



void consumer(BlockingQueue<std::string>* blkQueue) {
    std::string data;
    std::this_thread::sleep_for(std::chrono::seconds(5));

    for (int i = 0; i < 4; ++i) {
        data = blkQueue->take();
        cout << "\tConsumer: " << data << endl;

        if (0 == i)
            std::this_thread::sleep_for(std::chrono::seconds(5));
    }
}



int main() {
    BlockingQueue<std::string> blkQueue(2); // capacity = 2

    auto thProducer = std::thread(producer, &blkQueue);
    auto thConsumer = std::thread(consumer, &blkQueue);

    thProducer.join();
    thConsumer.join();

    return 0;
}

File: csharp/exercise/exer06b01-blocking-queue.cs
/*
 * BLOCKING QUEUE IMPLEMENTATION
 * Version B01: General blocking queues
 *              Underlying mechanism: Semaphores
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer06B01 : IRunnable
{
    public void run()
    {
        var queue = new MyBlockingQueue<string>(2); // capacity = 2
        new Thread(() => producer(queue)).Start();
        new Thread(() => consumer(queue)).Start();
    }


    private void producer(MyBlockingQueue<string> queue)
    {
        string[] arr = { "nice", "to", "meet", "you" };

        foreach (var value in arr)
        {
            Console.WriteLine($"Producer: {value}");
            queue.put(value);
            Console.WriteLine($"Producer: {value}\t\t\t[done]");
        }
    }


    private void consumer(MyBlockingQueue<string> queue)
    {
        string data;
        Thread.Sleep(5000);

        for (int i = 0; i < 4; ++i)
        {
            data = queue.take();
            Console.WriteLine($"\tConsumer: {data}");

            if (0 == i)
                Thread.Sleep(5000);
        }
    }


    //////////////////////////////////////////////


    class MyBlockingQueue<T>
    {
        private Semaphore semRemain = null;
        private Semaphore semFill = null;

        private int capacity = 0;
        private Queue<T> queue = null;


        public MyBlockingQueue(int capacity)
        {
            if (capacity <= 0)
                throw new ArgumentException("capacity must be a positive integer");

            semRemain = new Semaphore(capacity, capacity);
            semFill = new Semaphore(0, capacity);

            this.capacity = capacity;
            queue = new Queue<T>();
        }


        public void put(T value)
        {
            semRemain.WaitOne();

            lock (queue)
            {
                queue.Enqueue(value);
            }

            semFill.Release();
        }


        public T take()
        {
            T result = default;
            semFill.WaitOne();

            lock (queue) {
                result = queue.Dequeue();
            }

            semRemain.Release();
            return result;
        }
    }
}
